src/Chapter3.hs

```diff
@@ -1006,6 +1114,155 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- newtype Attack = Attack { attackValue :: Int } deriving (Show)
+-- newtype Health = Health { healthValue :: Int } deriving (Show)
+
+data KnightAction =
+  KnightAttack |
+  KnightDrinkHealthPotion Health |
+  KnightIncreaseDefenseSpell Defense
+  deriving (Show)
+
+-- Got conflicts
+data FKnight = FKnight
+  { kHealth :: Health
+  , kAttack :: Attack
+  , kDefense :: Defense
+  , kActions :: [KnightAction]
+  } deriving (Show)
+
+data MonsterAction =
+  MonsterAttack |
+  MonsterRunAway
+  deriving (Show)
+
+data FMonster = FMonster
+  { mHealth :: Health
+  , mAttack :: Attack
+  , mActions :: [MonsterAction]
+  } deriving (Show)
+
+class Fighter a where
+  calculateDamage :: Attack -> a -> Damage
+  calculateHit :: Damage -> a -> Maybe a
+  calculateAttack :: Attack -> a -> Maybe a
+  calculateAttack attack fighter = calculateHit (calculateDamage attack fighter) fighter
+
+damagedHealth :: Damage -> Health -> Health
+damagedHealth d h = Health (healthValue h - damageValue d)
+
+instance Fighter FKnight where
+  calculateDamage :: Attack -> FKnight -> Damage
+  calculateDamage a k = Damage (max 0 (attackValue a - defenseValue (kDefense k)))
+
+  calculateHit :: Damage -> FKnight -> Maybe FKnight
+  calculateHit d k
+    | healthValue (kHealth kAfter) <= 0 = Nothing
+    | otherwise = Just kAfter
+    where
+      kAfter = k {kHealth = damagedHealth d (kHealth k)}
+
+instance Fighter FMonster where
+  calculateDamage :: Attack -> FMonster -> Damage
+  calculateDamage a _ = Damage (attackValue a)
+
+  calculateHit :: Damage -> FMonster -> Maybe FMonster
+  calculateHit d m
+    | healthValue (mHealth mAfter) <= 0 = Nothing
+    | otherwise = Just mAfter
+    where
+      mAfter = m {mHealth = damagedHealth d (mHealth m)}
+
+
+-- How do I import that?
```

Comment:
this data type is not in the standard library.
I would suggest to create a custom type here, and you can even use better names that would more describe the purpose 🙂